### PR TITLE
functions to translate sample sheet in API

### LINKED ISSUE #2975: Command to translate a v1 sample sheet to a v2
## Description

Some v1 sample sheets can't be regenerated as v2 through the available commands in cg because their samples are not in LIMS. Therefore we need a way of reading the v1 sample sheet and translating it into a v2 sample sheet.

To generate a v2 sample sheet we need:
- Lane -> Present in the v1 sample sheet ✅ 
- Sample id -> Present in the v1 sample sheet ✅ 
- Indexes -> Present in the v1 sample sheet ✅  
- Index settings -> As these flow cells are old (HiSeq), they all have Non-reverse complement index settings ✅ 
- Read 1 and Read 2 run cycles ❌ 
- Index 1 and Index 2 run cycles ❌ -> Could be inferred from the index lengths?

The missing information is present in the RunParameters file. However, it is possible that this file is not available for some flow cells.

## Suggested solution
- [x] Implement the following in code
```mermaid
graph TD;
    v1_sample_sheet_file(("v1 Sample Sheet File"));
    Original_content["Original Content"];
    LIMS_samples["LIMS Samples"];
    Sample_sheet_creator["Sample Sheet Creator"];
    New_content["New Content"];
    Validated_content["Validated Content"];
    v2_sample_sheet_file(("v2 Sample Sheet File"));
    Flow_cell_path(("Flow Cell directory<br>with RunParameters<br>file"));
    FlowCell_Directory_Data["FlowCell Directory Data"];
    
    v1_sample_sheet_file -->|ReadFile.get_content_from_file| Original_content;
    Original_content -->|get_flow_cell_samples_from_content| LIMS_samples;
    LIMS_samples -->|parameter to| Sample_sheet_creator;
    Flow_cell_path -->|parameter to| FlowCell_Directory_Data;
    FlowCell_Directory_Data -->|parameter to| Sample_sheet_creator;
    Sample_sheet_creator -->|construct_sample_sheet| New_content;
    New_content -->|validate_sample_sheet_from_content| Validated_content;
    Validated_content -->|WriteStream.write_stream_from_content| v2_sample_sheet_file;

```

- [ ] See how to proceed with flow cells without RunParameters
- [ ] Do we need a CLI command? Or should the translation be set in the automation inside `cg demultiplex samplesheet create`?

## This can be closed when

There is a way to translate a V1 sample sheet to v2.

## Blocked by

- Issue of the old sample ids, see below

### PR DESCRIPTION
**Pull request by diitaz93**
## Description

Closes #2975 

### Added

-

### Changed

-

### Fixed

-

### How to prepare for test

- [ ] Ssh to relevant server (depending on type of change)
- [ ] Use stage: `us`
- [ ] Paxa the environment: `paxa`
- [ ] Install on stage (example for Hasta):
    ```shell
    bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_cg -t cg -b [THIS-BRANCH-NAME] -a
    ```

### How to test

- [ ] Do ...

### Expected test outcome

- [ ] Check that ...
- [ ] Take a screenshot and attach or copy/paste the output.

## Review

- [ ] Tests executed by
- [ ] "Merge and deploy" approved by
Thanks for filling in who performed the code review and the test!

### This [version](https://semver.org/) is a

- [ ] **MAJOR** - when you make incompatible API changes
- [ ] **MINOR** - when you add functionality in a backwards compatible manner
- [ ] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

## Implementation Plan

- [ ] Document in ...
- [ ] Deploy this branch on ...
- [ ] Inform to ...

### CODE REVIEW

**Comment by diitaz93** [comment]
## Tests on stage
New command exists
```shell
$ cg demultiplex samplesheet translate --help
Running cg demultiplex.
Usage: cg demultiplex samplesheet translate [OPTIONS] FLOW_CELL_NAME

  Translate a sample sheet from Bcl2Fastq to BclConvert format. 'flow-cell-
  name' is the flow cell run directory name, e.g.
  '181005_D00410_0735_BHM2LNBCX2' with a Bcl2Fastq sample sheet in it.

Options:
  -d, --dry-run  Runs the command without making any changes
  --help         Show this message and exit.
```
Translatable sample sheet
```shell
$ cg demultiplex samplesheet translate 170517_ST-E00266_0210_BHJCFFALXX --dry-run
Running cg demultiplex.
Translating Bcl2Fastq sample sheet for flow cell 170517_ST-E00266_0210_BHJCFFALXX
Called undefined __fields__ on HousekeeperAPI, please wrap
Sample sheet was generated for BCL2FASTQ
Removing of single index samples is not required for V2 sample sheet
Updating barcode mismatch values for samples in lane 2
Updating barcode mismatch values for samples in lane 3
Updating barcode mismatch values for samples in lane 4
Updating barcode mismatch values for samples in lane 5
Updating barcode mismatch values for samples in lane 6
Updating barcode mismatch values for samples in lane 7
Updating barcode mismatch values for samples in lane 8
Creating sample sheet content
Samplesheet passed BCLConvert validation
<content ommited>
```

**Comment by diitaz93** [comment]
## Tests on stage
### Flow cell without sample sheet
```shell
$ cg demultiplex samplesheet translate 170517_ST-E00266_0210_BHJCFFALXX
Running cg demultiplex.
Translating Bcl2Fastq sample sheet for flow cell 170517_ST-E00266_0210_BHJCFFALXX
Called undefined __fields__ on HousekeeperAPI, please wrap
Sample sheet for flow cell 170517_ST-E00266_0210_BHJCFFALXX does not exist <-- PRINTED IN RED
Traceback (most recent call last):
  File "/home/proj/stage/bin/miniconda3/envs/S_cg/bin/cg", line 8, in <module>
    sys.exit(base())
             ^^^^^^
  File "/home/proj/stage/bin/miniconda3/envs/S_cg/lib/python3.11/site-packages/click/core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/proj/stage/bin/miniconda3/envs/S_cg/lib/python3.11/site-packages/click/core.py", line 1078, in main
    rv = self.invoke(ctx)
         ^^^^^^^^^^^^^^^^
  File "/home/proj/stage/bin/miniconda3/envs/S_cg/lib/python3.11/site-packages/click/core.py", line 1688, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/proj/stage/bin/miniconda3/envs/S_cg/lib/python3.11/site-packages/click/core.py", line 1688, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/proj/stage/bin/miniconda3/envs/S_cg/lib/python3.11/site-packages/click/core.py", line 1688, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/proj/stage/bin/miniconda3/envs/S_cg/lib/python3.11/site-packages/click/core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/proj/stage/bin/miniconda3/envs/S_cg/lib/python3.11/site-packages/click/core.py", line 783, in invoke
    return __callback(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/proj/stage/bin/miniconda3/envs/S_cg/lib/python3.11/site-packages/click/decorators.py", line 45, in new_func
    return f(get_current_context().obj, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/proj/stage/bin/miniconda3/envs/S_cg/lib/python3.11/site-packages/cg/cli/demultiplex/sample_sheet.py", line 51, in translate_sample_sheet
    sample_sheet_api.translate_sample_sheet(flow_cell_name=flow_cell_name)
  File "/home/proj/stage/bin/miniconda3/envs/S_cg/lib/python3.11/site-packages/cg/apps/demultiplex/sample_sheet/api.py", line 142, in translate_sample_sheet
    raise SampleSheetError("Could not translate sample sheet")
cg.exc.SampleSheetError: Could not translate sample sheet
θ63° [js.diazboada@hasta:/home/proj/stage/sequencing_data/illumina/flow_cells] [S_base] 23s 1 $ 
```
### flow cell without run parameters
```shell
$ cg demultiplex samplesheet translate 170517_ST-E00266_0210_BHJCFFALXX
Running cg demultiplex.
Translating Bcl2Fastq sample sheet for flow cell 170517_ST-E00266_0210_BHJCFFALXX
Called undefined __fields__ on HousekeeperAPI, please wrap
No run parameters file found in flow cell /home/proj/stage/sequencing_data/illumina/flow_cells/170517_ST-E00266_0210_BHJCFFALXX <-- PRINTED IN RED
Run parameters file for flow cell 170517_ST-E00266_0210_BHJCFFALXX does not exist <-- PRINTED IN RED
Traceback (most recent call last):
  File "/home/proj/stage/bin/miniconda3/envs/S_cg/bin/cg", line 8, in <module>
    sys.exit(base())
             ^^^^^^
  File "/home/proj/stage/bin/miniconda3/envs/S_cg/lib/python3.11/site-packages/click/core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/proj/stage/bin/miniconda3/envs/S_cg/lib/python3.11/site-packages/click/core.py", line 1078, in main
    rv = self.invoke(ctx)
         ^^^^^^^^^^^^^^^^
  File "/home/proj/stage/bin/miniconda3/envs/S_cg/lib/python3.11/site-packages/click/core.py", line 1688, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/proj/stage/bin/miniconda3/envs/S_cg/lib/python3.11/site-packages/click/core.py", line 1688, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/proj/stage/bin/miniconda3/envs/S_cg/lib/python3.11/site-packages/click/core.py", line 1688, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/proj/stage/bin/miniconda3/envs/S_cg/lib/python3.11/site-packages/click/core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/proj/stage/bin/miniconda3/envs/S_cg/lib/python3.11/site-packages/click/core.py", line 783, in invoke
    return __callback(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/proj/stage/bin/miniconda3/envs/S_cg/lib/python3.11/site-packages/click/decorators.py", line 45, in new_func
    return f(get_current_context().obj, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/proj/stage/bin/miniconda3/envs/S_cg/lib/python3.11/site-packages/cg/cli/demultiplex/sample_sheet.py", line 51, in translate_sample_sheet
    sample_sheet_api.translate_sample_sheet(flow_cell_name=flow_cell_name)
  File "/home/proj/stage/bin/miniconda3/envs/S_cg/lib/python3.11/site-packages/cg/apps/demultiplex/sample_sheet/api.py", line 142, in translate_sample_sheet
    raise SampleSheetError("Could not translate sample sheet")
cg.exc.SampleSheetError: Could not translate sample sheet
θ61° [js.diazboada@hasta:/home/ … cing_data/illumina/flow_cells/170517_ST-E00266_0210_BHJCFFALXX] [S_base] 24s 1 $
```
### BCLConvert sample sheet
```shell
$ cg demultiplex samplesheet translate 170517_ST-E00266_0210_BHJCFFALXX
Running cg demultiplex.
Translating Bcl2Fastq sample sheet for flow cell 170517_ST-E00266_0210_BHJCFFALXX
Called undefined __fields__ on HousekeeperAPI, please wrap
Sample sheet was generated for BCL Convert
Sample sheet for flow cell 170517_ST-E00266_0210_BHJCFFALXX is not a Bcl2Fastq sample sheet <-- PRINTED IN RED
Traceback (most recent call last):
  File "/home/proj/stage/bin/miniconda3/envs/S_cg/bin/cg", line 8, in <module>
    sys.exit(base())
             ^^^^^^
  File "/home/proj/stage/bin/miniconda3/envs/S_cg/lib/python3.11/site-packages/click/core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/proj/stage/bin/miniconda3/envs/S_cg/lib/python3.11/site-packages/click/core.py", line 1078, in main
    rv = self.invoke(ctx)
         ^^^^^^^^^^^^^^^^
  File "/home/proj/stage/bin/miniconda3/envs/S_cg/lib/python3.11/site-packages/click/core.py", line 1688, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/proj/stage/bin/miniconda3/envs/S_cg/lib/python3.11/site-packages/click/core.py", line 1688, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/proj/stage/bin/miniconda3/envs/S_cg/lib/python3.11/site-packages/click/core.py", line 1688, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/proj/stage/bin/miniconda3/envs/S_cg/lib/python3.11/site-packages/click/core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/proj/stage/bin/miniconda3/envs/S_cg/lib/python3.11/site-packages/click/core.py", line 783, in invoke
    return __callback(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/proj/stage/bin/miniconda3/envs/S_cg/lib/python3.11/site-packages/click/decorators.py", line 45, in new_func
    return f(get_current_context().obj, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/proj/stage/bin/miniconda3/envs/S_cg/lib/python3.11/site-packages/cg/cli/demultiplex/sample_sheet.py", line 51, in translate_sample_sheet
    sample_sheet_api.translate_sample_sheet(flow_cell_name=flow_cell_name)
  File "/home/proj/stage/bin/miniconda3/envs/S_cg/lib/python3.11/site-packages/cg/apps/demultiplex/sample_sheet/api.py", line 142, in translate_sample_sheet
    raise SampleSheetError("Could not translate sample sheet")
cg.exc.SampleSheetError: Could not translate sample sheet
θ61° [js.diazboada@hasta:/home/ … cing_data/illumina/flow_cells/170517_ST-E00266_0210_BHJCFFALXX] [S_base] 23s 1 $ 
```

File: cg/apps/demultiplex/sample_sheet/api.py [hunk 1]
```diff
@@ -80,6 +94,78 @@ def validate_sample_sheet(
...
+        flow_cell: FlowCellDirectoryData = self._get_flow_cell(
+            flow_cell_name=flow_cell_name, bcl_converter=BclConverter.BCLCONVERT
+        )
+        if not self._is_sample_sheet_from_flow_cell_translatable(flow_cell):
+            raise SampleSheetError("Could not translate sample sheet")
+        old_content: list[list[str]] = self._replace_sample_header(
+            ReadFile.get_content_from_file(
+                file_format=FileFormat.CSV, file_path=flow_cell.sample_sheet_path
+            )
+        )
```
Comment (Vince-janv) [inline cg/apps/demultiplex/sample_sheet/api.py:147]: Why is the old_content retrieved by a function called `_replace_sample_header`? It seems to indicate a change in content

File: cg/apps/demultiplex/sample_sheet/read_sample_sheet.py [hunk 2]
```diff
@@ -34,7 +34,9 @@ def validate_samples_unique_per_lane(samples: list[FlowCellSample]) -> None:
         validate_samples_are_unique(samples=lane_samples)
 
 
-def get_sample_type_from_content(sample_sheet_content: list[list[str]]) -> Type[FlowCellSample]:
+def get_sample_type_from_content(
+    sample_sheet_content: list[list[str]],
+) -> Type[FlowCellSampleBcl2Fastq | FlowCellSampleBCLConvert]:
```
Comment (Vince-janv) [inline cg/apps/demultiplex/sample_sheet/read_sample_sheet.py:39]: Why is this changes? Are these not child classes?

File: cg/apps/demultiplex/sample_sheet/api.py [hunk 3]
```diff
@@ -80,6 +94,78 @@ def validate_sample_sheet(
...
+    @staticmethod
+    def _is_sample_sheet_from_flow_cell_translatable(flow_cell: FlowCellDirectoryData) -> bool:
+        """
+        Determine if the sample sheet from the flow cell directory is translatable to BCLConvert.
+        """
+        try:
+            flow_cell.run_parameters_path.exists()
+        except FlowCellError:
+            LOG.error(f"Run parameters file for flow cell {flow_cell.full_name} does not exist")
+            return False
```
Comment (Vince-janv) [inline cg/apps/demultiplex/sample_sheet/api.py:106]: Exists does not return an error it returns a bool I think this is what you want, maybe it can be incorporated into an `necessary_files_exist()` method together with the one below
```suggestion
        if not flow_cell.run_parameters_path.exists()
            LOG.error(f"Run parameters file for flow cell {flow_cell.full_name} does not exist")
            return False
```

File: cg/apps/demultiplex/sample_sheet/api.py [hunk 4]
```diff
@@ -80,6 +94,78 @@ def validate_sample_sheet(
...
+        if not flow_cell.sample_sheet_path.exists():
+            LOG.error(f"Sample sheet for flow cell {flow_cell.full_name} does not exist")
+            return False
+        sample_sheet_content: list[list[str]] = ReadFile.get_content_from_file(
+            file_format=FileFormat.CSV, file_path=flow_cell.sample_sheet_path
+        )
+        if get_sample_type_from_content(sample_sheet_content) is not FlowCellSampleBcl2Fastq:
```
Comment (Vince-janv) [inline cg/apps/demultiplex/sample_sheet/api.py:113]: How could this happen?

**Review by Vince-janv — COMMENTED** [review]
Nice, I think this is exactly the functionality we want 🙌  Some comments to fix

File: cg/apps/demultiplex/sample_sheet/read_sample_sheet.py (hunk 2 above)
Comment (diitaz93) [reply cg/apps/demultiplex/sample_sheet/read_sample_sheet.py:39]: Just to make Pycharm not complain. Should I revert this? @Vince-janv

File: cg/apps/demultiplex/sample_sheet/api.py (hunk 3 above)
Comment (diitaz93) [reply cg/apps/demultiplex/sample_sheet/api.py:106]: Unfortunately `flow_cell.run_parameters_path` raises an error if the run parameters does not exists, so the `try-except` clause is needed here. I think that maybe this could not be the best behaviour for this function but I think that changing this behaviour would be out of the scope of this PR. 
I could however remove the `.exists()` and give it a bit more clarity to this part, what do you think? @Vince-janv

File: cg/apps/demultiplex/sample_sheet/api.py (hunk 4 above)
Comment (diitaz93) [reply cg/apps/demultiplex/sample_sheet/api.py:113]: This is when you try to translate to BCLConvert a sample sheet that is already in BCLConvert format

File: cg/apps/demultiplex/sample_sheet/read_sample_sheet.py (hunk 2 above)
Comment (Vince-janv) [reply cg/apps/demultiplex/sample_sheet/read_sample_sheet.py:39]: What is Pycharm complaining about?

File: cg/apps/demultiplex/sample_sheet/api.py (hunk 3 above)
Comment (Vince-janv) [reply cg/apps/demultiplex/sample_sheet/api.py:106]: hmm. Does it make sense thet the property `run_parameters_path` raises an error? Maybe it can just return `None` if there is not file. That would make more sense to me. What do you think?

File: cg/apps/demultiplex/sample_sheet/read_sample_sheet.py (hunk 2 above)
Comment (diitaz93) [reply cg/apps/demultiplex/sample_sheet/read_sample_sheet.py:39]: Just type hint warnings

File: cg/apps/demultiplex/sample_sheet/api.py (hunk 3 above)
Comment (diitaz93) [reply cg/apps/demultiplex/sample_sheet/api.py:106]: It makes sense that it returns None instead of error. I would have to see if fixing that would extend tremendously the scope of this PR, as this attribute is used in many places

File: cg/apps/demultiplex/sample_sheet/read_sample_sheet.py (hunk 2 above)
Comment (diitaz93) [reply cg/apps/demultiplex/sample_sheet/read_sample_sheet.py:39]: resolved in person

**Review by henrikstranneheim — APPROVED** [review]
Nice, 👍

File: tests/apps/demultiplex/test_translate_sample_sheet.py [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+    """Test that the header is replaced correctly for a Bcl2Fastq sample sheet."""
+    # GIVEN a sample sheet api
+    api: SampleSheetAPI = sample_sheet_context.sample_sheet_api
+
+    # GIVEN a Bcl2Fastq sample sheet content
+
+    # WHEN replacing the header
+    new_content: list[list[str]] = api._replace_sample_header(sample_sheet_bcl2fastq_data_header)
+
+    # THEN the new header is correct
```
[the remark this answers is not in the document]
Comment (diitaz93) [reply tests/apps/demultiplex/test_translate_sample_sheet.py:111]: Yes, that the Bcl2Fastq header `SampleID` is replaced by `Sample_ID`. I will change the statement to mention this

File: cg/apps/demultiplex/sample_sheet/api.py [hunk 6]
```diff
@@ -80,6 +94,78 @@ def validate_sample_sheet(
...
+            raise SampleSheetError("Could not translate sample sheet")
+        old_content: list[list[str]] = self._replace_sample_header(
+            ReadFile.get_content_from_file(
+                file_format=FileFormat.CSV, file_path=flow_cell.sample_sheet_path
+            )
+        )
+        flow_cell_samples: list[FlowCellSampleBCLConvert] = get_flow_cell_samples_from_content(
+            sample_sheet_content=old_content, sample_type=FlowCellSampleBCLConvert
+        )
+        creator = SampleSheetCreatorBCLConvert(flow_cell=flow_cell, lims_samples=flow_cell_samples)
```
[the remark this answers is not in the document]
Comment (diitaz93) [reply cg/apps/demultiplex/sample_sheet/api.py:151]: fixed

File: cg/apps/demultiplex/sample_sheet/api.py (hunk 1 above)
Comment (diitaz93) [reply cg/apps/demultiplex/sample_sheet/api.py]: fixed

File: cg/apps/demultiplex/sample_sheet/api.py [hunk 7]
```diff
@@ -80,6 +94,78 @@ def validate_sample_sheet(
...
+            LOG.error(
+                f"Sample sheet for flow cell {flow_cell.full_name} is not a Bcl2Fastq sample sheet"
+            )
+            return False
+        return True
+
+    @staticmethod
+    def _replace_sample_header(sample_sheet_content: list[list[str]]) -> list[list[str]]:
+        """
+        Replace the old sample id header in the Bcl2Fastq sample sheet content with the new one.
```
[the remark this answers is not in the document]
Comment (diitaz93) [reply cg/apps/demultiplex/sample_sheet/api.py:123]: fixed

File: cg/apps/demultiplex/sample_sheet/api.py (hunk 3 above)
Comment (diitaz93) [reply cg/apps/demultiplex/sample_sheet/api.py:106]: will be solved in another PR, described in #3083

File: cg/apps/demultiplex/sample_sheet/api.py (hunk 4 above)
Comment (diitaz93) [reply cg/apps/demultiplex/sample_sheet/api.py:116]: done! created smaller functions

File: cg/apps/demultiplex/sample_sheet/api.py [hunk 8]
```diff
@@ -80,6 +94,96 @@ def validate_sample_sheet(
...
+        self, flow_cell: FlowCellDirectoryData
+    ) -> bool:
+        """
+        Determine if the sample sheet from the flow cell directory is translatable to BCLConvert.
+        """
+        if not self._are_necessary_files_in_flow_cell(flow_cell):
+            return False
+        if not self._is_sample_sheet_bcl2fastq(flow_cell):
+            return False
+        return True
```
Comment (Vince-janv) [inline cg/apps/demultiplex/sample_sheet/api.py:133]: I dislike functions that have multiple return statements. I think they become hard to follow.
I think this is cleaner
```suggestion
       return all(self._are_necessary_files_in_flow_cell(flow_cell) and self._is_sample_sheet_bcl2fastq(flow_cell))
```

File: cg/apps/demultiplex/sample_sheet/api.py [hunk 9]
```diff
@@ -80,6 +94,96 @@ def validate_sample_sheet(
...
+                return sample_sheet_content
+        raise SampleSheetError("Could not find data header in sample sheet")
+
+    def translate_sample_sheet(self, flow_cell_name: str) -> None:
+        """Translate a Bcl2Fastq sample sheet to a BCLConvert sample sheet."""
+        flow_cell: FlowCellDirectoryData = self._get_flow_cell(
+            flow_cell_name=flow_cell_name, bcl_converter=BclConverter.BCLCONVERT
+        )
+        if not self._is_sample_sheet_from_flow_cell_translatable(flow_cell):
+            raise SampleSheetError("Could not translate sample sheet")
```
Comment (Vince-janv) [inline cg/apps/demultiplex/sample_sheet/api.py:158]: Much cleaner ⭐

**Review by Vince-janv — APPROVED** [review]
Great! Just one small comment 👍

File: cg/apps/demultiplex/sample_sheet/api.py (hunk 8 above)
Comment (diitaz93) [reply cg/apps/demultiplex/sample_sheet/api.py]: fixed
